framework/decode/vulkan_virtual_swapchain.cpp

```diff
@@ -416,196 +534,108 @@ VkResult VulkanVirtualSwapchain::QueuePresentKHR(PFN_vkQueuePresentKHR
-    final_barrier_virtual_image.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
-    final_barrier_virtual_image.dstAccessMask = VK_ACCESS_MEMORY_READ_BIT;
-    final_barrier_virtual_image.oldLayout     = VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL;
-    final_barrier_virtual_image.newLayout     = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;
-
-    initial_barrier_swapchain_image               = initial_barrier_virtual_image;
-    initial_barrier_swapchain_image.dstAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
-    initial_barrier_swapchain_image.newLayout     = VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
-
-    final_barrier_swapchain_image           = final_barrier_virtual_image;
-    final_barrier_swapchain_image.oldLayout = VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
-
-    VkImageSubresourceLayers subresource = {
-        VK_IMAGE_ASPECT_COLOR_BIT,
-        0,
-        0,
-        0,
-    };
-
-    VkOffset3D offsets[2] = {
-        {
-            0,
-            0,
-            0,
-        },
-        {
-            0,
-            0,
-            1,
-        },
-    };
+    VkCommandBufferBeginInfo begin_info  = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO, nullptr, 0, nullptr };
+    VkImageSubresourceLayers subresource = { VK_IMAGE_ASPECT_COLOR_BIT, 0, 0, 0 };
+    VkOffset3D               offset      = { 0, 0, 0 };
 
-    auto                     length = present_info->swapchainCount;
-    std::vector<VkSemaphore> semaphores;
+    auto                     swapchainCount = present_info->swapchainCount;
+    std::vector<VkSemaphore> present_wait_semaphores;
 
-    for (uint32_t i = 0; i < length; ++i)
+    std::vector<VkPipelineStageFlags> wait_stages;
+    wait_stages.resize(present_info->waitSemaphoreCount);
+    for (uint32_t j = 0; j < present_info->waitSemaphoreCount; ++j)
     {
+        wait_stages.push_back(VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT);
+    }
+
+    for (uint32_t i = 0; i < swapchainCount; ++i)
+    {
+        const auto* swapchain_info      = swapchain_infos[i];
         uint32_t    capture_image_index = capture_image_indices[i];
         uint32_t    replay_image_index  = present_info->pImageIndices[i];
-        const auto* swapchain_info      = swapchain_infos[i];
-        const auto& virtual_image       = swapchain_info->virtual_images[capture_image_index];
-        const auto& swapchain_image     = swapchain_info->swapchain_images[replay_image_index];
+        const auto& virtual_image       = swapchain_info->virtual_swapchain_images[capture_image_index];
+        const auto& replay_image        = swapchain_info->replay_swapchain_images[replay_image_index];
+
+        // Use a command buffer and semaphore from the same queue index
+        assert(swapchain_info->device_info->queue_family_indices.find(queue) !=
+               swapchain_info->device_info->queue_family_indices.end());
+        uint32_t queue_family_index = swapchain_info->device_info->queue_family_indices[queue];
```

Comment:
If queue_family_index saves in QueueInfo, we don't need to find the Queue in DeviceInfo here.

Reply:
See above.

Reply:
Done.